outlines/fsm/fsm.py

```diff
@@ -119,9 +140,63 @@ def __init__(self, regex_string: str, tokenizer: "Tokenizer"):
         self.final_states = regex_fsm.finals | {
             -1
         }  # Include the EOS token in final states
-        self.vocabulary = tokenizer.vocabulary.values()
+        self.tokenizer = tokenizer
+        self.vocabulary = tokenizer.vocabulary
         self.end_token_id = tokenizer.eos_token_id
 
+    def align_prompt_tokens(self, prompt: str) -> str:
+        """Remove the last token from the prompt and update the states_to_token_maps accordingly"""
+        token_ids, _ = self.tokenizer.encode(prompt)
+        last_token_id = int(token_ids[0][-1])
+        last_token_text = self.tokenizer.decode([last_token_id])[0]
+        last_token_length = len(last_token_text)
+        # select the tokens that start with the text removed from the prompt
+        crossing_tokens = {
+            token: text
+            for text, token in self.vocabulary.items()
+            if text.startswith(last_token_text)
+        }
```

Comment:
The section above would be re-used for the 3 FSMs. This could be a sign that the code is not located where it should be.